server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,850 @@
+                    }
+                } catch (Throwable t) {
+                    logger.warn("Failed to delete the old primary repository: {}", repoDir);
+                }
+            });
+            logger.info("Promotion is done for {}/{}. {} is now the primary.",
+                        parent.name(), originalRepoName, secondaryRepo.repoDir());
+        } finally {
+            writeUnLock();
+        }
+
+        createSecondaryRepo();
+    }
+
+    private static boolean deleteDirectory(File dir) throws IOException {
+        try (Stream<Path> walk = Files.walk(dir.toPath())) {
+            return walk.sorted(Comparator.reverseOrder())
+                       .map(Path::toFile)
+                       .map(File::delete)
+                       // Return false if it fails to delete a file.
+                       .reduce(true, (a, b) -> a && b);
+        }
+    }
+
+    private void createSecondaryRepo() {
+        try {
+            writeLock();
+            final Revision headRevision = this.headRevision;
+            isCreatingSecondaryRepo = true;
+            writeUnLock();
+
+            logger.info("Creating the secondary repository in {}/{} with the head revision: {}.",
+                        parent.name(), originalRepoName, headRevision);
+            final Map<String, Entry<?>> entries = find(headRevision, ALL_PATH, ImmutableMap.of()).join();
+            final List<Change<?>> changes = toChanges(entries);
+            final File secondaryRepoDir = repoMetadata.secondaryRepoDir();
+            final InternalRepository secondaryRepo =
+                    InternalRepository.of(parent, originalRepoName, secondaryRepoDir, headRevision,
+                                          creationTimeMillis, author, changes);
+            writeLock();
+            try {
+                if (laggedCommitsForSecondary.isEmpty()) {
+                    // There were no commits after creating the secondary repo
+                    // so the headRevision hasn't changed.
+                    assert headRevision == this.headRevision;
+                } else {
+                    // We should catch up.
+                    for (LaggedCommit c : laggedCommitsForSecondary) {
+                        secondaryRepo.commit(c.revision, c.revision.forward(1), c.commitTimeMillis, c.author,
+                                             c.summary, c.detail, c.markup, c.changes, c.allowEmpty);
+                    }
+                    laggedCommitsForSecondary.clear();
+                }
+                isCreatingSecondaryRepo = false;
+                this.secondaryRepo = secondaryRepo;
+            } finally {
+                writeUnLock();
+            }
+            logger.info("The secondary repository {} is created in {}/{}.",
+                        secondaryRepoDir.getName(), parent.name(), originalRepoName, headRevision);
+        } catch (Throwable t) {
+            logger.warn("Failed to create the secondary repository", t);
```

Comment:
Don't we need to throw an exception and stop replication rather than catching it if an exception is raised while creating a secondary repository?
Additionally, we might need to delete the incomplete secondary repository just in case only half of `LaggedCommit`s is saved.

Reply:
> Additionally, we might need to delete the incomplete secondary repository just in case only half of LaggedCommits is saved.

👍 

> Don't we need to throw an exception and stop replication rather than catching it if an exception is raised while creating a secondary repository?

I thought we can continue to serve because we can serve the request without a second repository.
The secondary repository does not have to be created in a timely manner and the plugin tries that in the next schedule. 🤔

Reply:
Should we need to set `isCreatingSecondaryRepo` to false and clear `laggedCommitsForSecondary` if there is an error while committing data to the secondary repository?